--
title: "ADSP WGS GWAS results"
author: "Xulong Wang"
date: "May 3, 2016"
output: pdf_document
---

```{r, eval = F}

setwd("/data/xwang/adsp3/glm")
setwd("/data/xwang/adsp3/glm2") # Binary

glm <- lapply(dir(), function(x) { load(x); fit })
lrt <- lapply(glm, function(x) {y = x[, "lp"]; y = y - min(y); names(y) = rownames(x); y})

glm <- as.data.frame(do.call(rbind, glm))
glm$P <- pnorm(abs(glm$p), sd = glm$se, lower.tail = F) * 2

#---

setwd("/data/xwang/adsp3/sampling")
setwd("/data/xwang/adsp3/sampling2") # Binary

mcmc <- lapply(dir(), function(x) { load(x); fit })
mcmc <- lapply(mcmc, function(x) { lapply(x, function(y) { sapply(y[all(! grepl("Error", y))], function(z) z["p", ]) }) })

mcmc <- lapply(mcmc, function(x) do.call(cbind, x))
mcmc <- t(do.call(cbind, mcmc[sapply(mcmc, function(x) dim(x)[1]) == 10]))

mcmc <- as.data.frame(mcmc)
mcmc$P <- pnorm(abs(unlist(mcmc$mean)), sd = unlist(mcmc$sd), lower.tail = F) * 2

save(mcmc, file = "/data/xwang/adsp3/sampling.rdt")

#---

library(ggplot2)
options(stringsAsFactors = F)

chrlen <- read.delim("~/Dropbox/GitHub/X/genomes/human.hg19.genome", header = F)
chrlen <- chrlen[match(paste0("chr", 1:22), chrlen$V1), ]
chrlen <- cumsum(as.numeric(chrlen$V2)) * 1e-6; names(chrlen) <- c(1:22)
chrmid <- diff(c(0, chrlen)) * 0.5 + c(0, chrlen[-length(chrlen)])

manhattan <- glm
manhattan <- mcmc

manhattan$CHR = as.numeric(gsub(":.*", "", rownames(manhattan)))
manhattan$POS = as.numeric(gsub(".*:(.*)_.*", "\\1", rownames(manhattan)))

manhattan$POS2 <- c(0, chrlen)[manhattan$CHR] + manhattan$POS * 1e-6
manhattan$col <- rep("o", nrow(manhattan))
manhattan$col[manhattan$CHR %% 2 == 1] <- "e"

manhattan$P <- -log10(manhattan$P)

png("~/Dropbox/GitHub/Adsp/new/manhattan_mcmc.png", width = 2e3, height = 1e3, res = 200)
png("~/Dropbox/GitHub/Adsp/new/manhattan_mcmc2.png", width = 2e3, height = 1e3, res = 200)

ggplot(manhattan, aes(x = POS2, y = P, color = col)) + 
  geom_point(alpha = 0.9) + geom_hline(yintercept = 7.3, color = "gold") +
  scale_x_continuous(breaks = chrmid, labels = names(chrlen)) + 
  scale_color_manual(values = c("dodgerblue3", "firebrick1")) + 
  theme_bw() + xlab("") + ylab("-log10(P)") + guides(shape = F, color = F) +
  theme(legend.key = element_blank())

dev.off()

```

#--- Significant variants by MCMC

```{r}

load("~/Dropbox/GitHub/Adsp/new/mdata.rdt")
load("~/Dropbox/GitHub/Adsp/new/mcmc_sig.rdt")

mcmc_sig = x[, mdata$ADSP.Sample.ID]
y = within(mdata, { geno = mcmc_sig[4, ] })

table(y$AD1, y$geno)
ggplot(y, aes(x = AD1, fill = as.factor(geno))) + geom_bar()

pdf("~/Dropbox/GitHub/Adsp/new/loci.pdf", width = 5, height = 5)
ggplot(y, aes(x = AD1, fill = as.factor(geno))) + 
  geom_bar(position = "fill") + theme_bw() + xlab("") + ylab("Proportion") 
dev.off()

```

### Consequences

```{r}

load("~/Desktop/GWAS/vepList.rdt"); vep = do.call(rbind, vepList)

glmm = glmm_p0.01[glmm_p0.01$P < 1e-4, ]

vep1 = vep[vep$X.Uploaded_variation %in% rownames(glmm), ]
vep1[vep1$Consequence == "missense_variant", ]

```

x = manhattan[which(manhattan$pse > 7.3), ]
x = x[x$chr != "19", ]
vep_input = data.frame(chr = x$chr, start = gsub("*.-", "", x$uid))
vep_input$end = vep_input$start
vep_input$allele = "A/T"
vep_input$strand = "+"

write.table(vep_input, file = "~/Dropbox/GitHub/glmm/vep_input.txt", quote = F, sep = "\t", row.names = F)

vep = read.table("./gwas/vep.txt", header = T)
(genes = unique(vep$SYMBOL))

```

### Compare the glmm and glmm_prior (sigma = 0.5)

```{r}

load("~/Dropbox/GitHub/glmm/gwas/glmm_p0.01.rdt")
load("~/Dropbox/GitHub/glmm/gwas/glmm_prior_p0.01.rdt")
load("~/Dropbox/GitHub/glmm/igap/igap.rdt")
load("~/Dropbox/GitHub/glmm/data/glmList.rdt")

glm_old = glmList$gwas[glmList$gwas$LOD > 15,]
glm_old = glm_old[glm_old$CHR %in% c(1:22), ]

table(glm_old$UID %in% rownames(glmm_p0.01))
glm_old$P = pchisq(glm_old$LOD, df = 1, lower.tail = F)

plot(-log10(glm_old$P), -log10(glmm_p0.01[glm_old$UID, "P"]))
cor(-log10(glm_old$P), -log10(glmm_p0.01[glm_old$UID, "P"]), method = "spearman")

glmm = glmm_p0.01[glmm_p0.01$P < 1e-7, ]

uid = rownames(glmm)[rownames(glmm) %in% rownames(glmm_prior_p0.01)]
glmm[uid, ]
glmm_prior_p0.01[uid, ]

```
